src/compiler/scala/tools/nsc/typechecker/Infer.scala

```diff
@@ -1515,6 +1514,8 @@ trait Infer extends Checkable {
     }
   }
 
+  private[this] object detectTopTypesCollector extends ContainsAnyCollector(AnyClass :: AnyValClass :: ObjectClass :: Nil)
+
```

Comment:
Missed opportunity to call it `ContainsAnyAnyCollector`.

Reply:
If you abstract out just the list of syms -- the other day I grepped to see if anything else could use it, but other checks are for Any and AnyVal; offhand, I'm not sure where Object comes in here -- then the check becomes
```
for (t <- targs) if (anysyms.contains(t.typeSymbol)) reporterr.warning()
```

Reply:
Not sure I understand...  could you expand what the full solution would be?

Reply:
I don't have a strong opinion about the micro-optimization, but if you extract
```
private[this] val anysyms = List(AnyClass,...)
```
then leave the innocent allocation `val collector = new ContainsAnyCollector(anysyms)` and replace the match with the for loop shown. Probably the match is faster than contains, but contains is easier to read. WDYT? Or actually, DWYW for do what pleases you.

Reply:
Oh, I see. You meant 7 lines down from the change, where the current code reads
```Scala
        targs.foreach(_.typeSymbol match {
          case sym @ (AnyClass | AnyValClass | ObjectClass) =>
            reporter.warning(fn.pos, s"a type was inferred to be `${sym.name}`; this may indicate a programming error.")
          case _ =>
```
To use instead 
```Scala
        for (t <- targs ; sym = t.typeSymbol ; if anysyms.contains(sym) 
            reporter.warning(fn.pos, s"a type was inferred to be `${sym.name}`; this may indicate a programming error.")
```

I am not sure if a call to `List.contains`, which needs to go through three nodes in memory, would be faster than a pattern match which comes down to three equality comparisons. On the other hand, it may make sense to specialise that `detectTopTypesCollector` to one that specifically checks for those three constant elements, not use the one intended for any list. I'll do that change.

Reply:
I also believe hoisting `List(AnyClass, AnyValClass, ObjectClass)` to a private field in `Infer` and otherwise leaving the code as is would be an appropriate change. This code path is only hit under `-Xlint`. That doesn't mean we shouldn't care about its performance, but as I've mentioned elsewhere, we should use profile results to motivate this style of change, especially when the refactoring adds lines of code or makes things a bit harder to grok.

Reply:
Noting that "only hit under `-Xlint`" means it should always be hit.